message_endpoint.go

```diff
@@ -0,0 +1,293 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package swim
+
+import (
+	"errors"
+
+	"github.com/DE-labtory/swim/pb"
+	"github.com/gogo/protobuf/proto"
+	"github.com/it-chain/iLogger"
+
+	"time"
+)
+
+var ErrSendTimeout = errors.New("Error send timeout")
+var ErrUnreachable = errors.New("Error this shouldn't reach")
+var ErrInvalidMessage = errors.New("Error invalid message")
+var ErrCallbackCollectIntervalNotSpecified = errors.New("Error callback collect interval should be specified")
+
+// sallback is called when target member sent back to local member a message
+// created field is for clean up the old callback
+type callback struct {
+	fn func(msg pb.Message)
+	created time.Time
+}
+
+// responseHandler manages callback functions
+type responseHandler struct {
+	callbacks map[string]callback
+	collectInterval time.Duration
+}
+
+func newResponseHandler(collectInterval time.Duration) *responseHandler {
+	h := &responseHandler{
+		callbacks: make(map[string]callback),
+		collectInterval: collectInterval,
+	}
+
+	go h.collectCallback()
+
+	return h
+}
+
+func (r *responseHandler) addCallback(seq string, cb callback) {
+	r.callbacks[seq] = cb
+}
+
+// Handle, each time other member sent back
+// a message, callback matching message's seq is called
+func (r *responseHandler) handle(msg pb.Message) {
+	seq := msg.Seq
+	cb, exist := r.callbacks[seq]
+
+	if exist == false {
+		iLogger.Panicf(nil, "Panic, no matching callback function")
+	}
+
+	cb.fn(msg)
+	delete(r.callbacks, seq)
+}
+
+func (r *responseHandler) hasCallback(seq string) bool {
+	for s := range r.callbacks {
+		if seq == s {
+			return true
+		}
+	}
+	return false
+}
+
+// collectCallback every time callbackCollectInterval expired clean up
+// the old (time.now - callback.created > time interval) callback delete from map
+// callbackCollectInterval specified in config
+func (r *responseHandler) collectCallback() {
+	timeout := r.collectInterval
+	T := time.NewTimer(timeout)
+
+	for {
+		select {
+		case <- T.C:
+			for seq, cb := range r.callbacks {
+				if time.Now().Sub(cb.created) > timeout {
+					delete(r.callbacks, seq)
+				}
+			}
+			T = time.NewTimer(timeout)
+		}
+	}
+}
+
+type MessageEndpointConfig struct {
+	EncryptionEnabled  bool
+	DefaultSendTimeout time.Duration
+
+	// callbackCollect Interval indicate time interval to clean up old
+	// callback function
+	CallbackCollectInterval time.Duration
+}
+
+// MessageEndpoint basically do receiving packet and determine
+// which logic executed based on the packet.
+type MessageEndpoint struct {
+	config         MessageEndpointConfig
+	transport      UDPTransport
+	messageHandler MessageHandler
+	awareness      *Awareness
+	resHandler     *responseHandler
+	quit           chan struct{}
```

Comment:
@hihiboss 5 hours ago  Member
resHandler is reference but Why messageHandler is not reference?!

  @zeroFruit
zeroFruit 36 minutes ago  • 
 Member
I coded like this because I think that each MessageEndpointshould reflect respondHandler's change. and I've remembered interface as pointer of struct, and MessageHandler is interface, resHandler struct. so I've tried to keep consistency